examples/siro_sandbox/server/server.py

```diff
@@ -0,0 +1,188 @@
+from .frequency_limiter import FrequencyLimiter
+
+# Boolean variable to indicate whether to use SSL
+use_ssl = False
+
+server_process = None
+exit_event = None
+
+
+def launch_server_process(interprocess_record):
+    # see multiprocessing_config to switch between real and dummy multiprocessing
+
+    global server_process
+    global exit_event
+
+    # multiprocessing.dummy.Process is actually a thread and requires special logic
+    # to terminate it.
+    exit_event = (
+        threading.Event() if multiprocessing_config.use_dummy else None
+    )
+
+    server_process = multiprocessing_config.Process(
+        target=server_main, args=(interprocess_record, exit_event)
+    )
+    server_process.start()
+
+
+def terminate_server_process():
+    global server_process
+    global exit_event
+    if multiprocessing_config.use_dummy:
+        if exit_event:
+            exit_event.set()
+            exit_event = None
+    else:
+        if server_process:
+            server_process.terminate()
+            server_process = None
+
+
+def create_ssl_context():
+    ssl_context = ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER)
+    certfile = "./examples/siro_sandbox/self_signed.pem"
+    keyfile = "./examples/siro_sandbox/private.key"
+    # see instructions for generating these files: https://docs.google.com/document/d/1hXuStZKNJafxLQVgl2zy2kyEf8Nke-bYogaPuHYT_M4/edit#bookmark=id.jva9nto0xpbe
+    assert os.path.exists(certfile) and os.path.exists(keyfile)
+    ssl_context.load_cert_chain(certfile=certfile, keyfile=keyfile)
+    return ssl_context
+
+
+class Server:
+    def __init__(self, interprocess_record, exit_event):
+        self._connected_clients = {}  # Dictionary to store connected clients
+
+        self._interprocess_record = interprocess_record
+
+        # Limit how many messages/sec we send. Note the current server implementation sends
+        # messages "one at a time" (waiting for confirmation of receipt from the
+        # remote client OS), so there is probably no risk of overwhelming the
+        # network connection bandwidth even when sending at a high rate.
+        max_send_rate = None  # or set None to not limit
+        self._send_frequency_limiter = FrequencyLimiter(max_send_rate)
+
+        self._exit_event = exit_event
+
+        self._hack_first_keyframe = None
+
+    async def send_keyframes(self, websocket):
+        did_send_hack_first_keyframe = False
+
+        while True:
+            if self._exit_event and self._exit_event.is_set():
+                break
+
+            # todo: refactor to support N clients
+            keyframes = self._interprocess_record.get_queued_keyframes()
+
+            if len(keyframes):
+                if not self._hack_first_keyframe:
+                    self._hack_first_keyframe = keyframes[0]
+                elif not did_send_hack_first_keyframe:
+                    # hack: resend first keyframe to new clients
+                    keyframes.insert(0, self._hack_first_keyframe)
```

Comment:
This is a hack which isn't working very well. Context: when a new client connects after the server has already been running a while, we can't simply start sending it incremental gfx-replay keyframes--the client won't get the instance creations that happened in the past. As a hack here, we hope/assume that all instance-creations happen on frame 0, and thus if we inject frame 0, maybe everything will work fine. This isn't robust, and so currently we don't really support clients connecting after the server has already started simulation. All this can be fixed by using our "gfx-replay keyframe consolidation" logic in Habitat-sim, which isn't properly exposed to Python yet.